### PR TITLE
Make magic tags behave the same in various locations

### PR DESCRIPTION
**Pull request by JoryHogeveen**
Currently every location where magic tags are used it behaved differently, not all options are always supported. This PR aims to improve this and support all tags on all locations.

- Pods templates
- Shortcodes (content and parameters)

This PR also adds `before` and `after` params to special magic tags. These are already supported for Pod related magic tags.
- `{@user.id,helper,before,after}`

Merging this PR will also merge #5713 since it's dependency.

Fixes: #5734 
Fixes: #5697 
Fixes: #4887

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.
- [ ] My code includes PHP Unit Tests (if applicable)

### CODE REVIEW

File: classes/Pods.php [hunk 1]
```diff
@@ -4175,6 +4175,16 @@ private function process_magic_tags( $tag ) {
...
+		// Process special magic tags but allow "empty" values for numbers.
+		if (
+			! $value
+			&& ! is_numeric( $value )
+			&& pods_shortcode_allow_evaluate_tags()
+			&& ! $this->fields( $field_name )
+		) {
+			$value = pods_evaluate_tag( implode( ',', $tag ) );
+		}
+
```
Comment (sc0ttkclark) [inline classes/Pods.php:4187]:
```suggestion
```

File: includes/general.php [hunk 2]
```diff
@@ -972,15 +972,15 @@ function pods_shortcode_run( $tags, $content = null ) {
 				$params['where'] = $tags['where'];
 
 				if ( pods_shortcode_allow_evaluate_tags() ) {
-					$params['where'] = pods_evaluate_tags( html_entity_decode( $params['where'] ), true, '""' );
+					$params['where'] = pods_evaluate_tags( html_entity_decode( $params['where'] ), true, '""', true );
```
Comment (sc0ttkclark) [inline includes/general.php:975]:
```suggestion
					$params['where'] = pods_evaluate_tags( html_entity_decode( $params['where'] ), true, '""', $pod );
```

File: includes/general.php [hunk 3]
```diff
@@ -972,15 +972,15 @@ function pods_shortcode_run( $tags, $content = null ) {
...
 				}
 			}
 
 			if ( 0 < strlen( $tags['having'] ) ) {
 				$params['having'] = $tags['having'];
 
 				if ( pods_shortcode_allow_evaluate_tags() ) {
-					$params['having'] = pods_evaluate_tags( html_entity_decode( $params['having'] ), true, '""' );
+					$params['having'] = pods_evaluate_tags( html_entity_decode( $params['having'] ), true, '""', true );
```
Comment (sc0ttkclark) [inline includes/general.php:983]:
```suggestion
					$params['having'] = pods_evaluate_tags( html_entity_decode( $params['having'] ), true, '""', $pod );
```

File: includes/general.php [hunk 4]
```diff
@@ -926,15 +926,15 @@ function pods_shortcode_run( $tags, $content = null ) {
 			$id = $tags['slug'];
 
 			if ( pods_shortcode_allow_evaluate_tags() ) {
-				$id = pods_evaluate_tags( $id, true );
+				$id = pods_evaluate_tags( $id, true, null, true );
```
Comment (sc0ttkclark) [inline includes/general.php:929]:
```suggestion
				$id = pods_evaluate_tags( $id, true );
```

File: includes/general.php [hunk 5]
```diff
@@ -926,15 +926,15 @@ function pods_shortcode_run( $tags, $content = null ) {
...
 			}
 		}
 
 		if ( ! empty( $tags['id'] ) ) {
 			$id = $tags['id'];
 
 			if ( pods_shortcode_allow_evaluate_tags() ) {
-				$id = pods_evaluate_tags( $id, true );
+				$id = pods_evaluate_tags( $id, true, null, true );
```
Comment (sc0ttkclark) [inline includes/general.php:937]:
```suggestion
				$id = pods_evaluate_tags( $id, true );
```

File: classes/Pods.php [hunk 6]
```diff
@@ -4175,16 +4175,6 @@ private function process_magic_tags( $tag ) {
...
-		// Process special magic tags but allow "empty" values for numbers.
-		if (
-			! $value
-			&& ! is_numeric( $value )
-			&& pods_shortcode_allow_evaluate_tags()
-			&& ! $this->fields( $field_name )
-		) {
-			$value = pods_evaluate_tag( implode( ',', $tag ) );
-		}
-
```
Comment (JoryHogeveen) [inline classes/Pods.php]: Why did you remove this?

File: includes/general.php [hunk 7]
```diff
@@ -972,15 +972,15 @@ function pods_shortcode_run( $tags, $content = null ) {
 				$params['where'] = $tags['where'];
 
 				if ( pods_shortcode_allow_evaluate_tags() ) {
-					$params['where'] = pods_evaluate_tags( html_entity_decode( $params['where'] ), true, '""', true );
+					$params['where'] = pods_evaluate_tags( html_entity_decode( $params['where'] ), true, '""', $pod );
```
Comment (JoryHogeveen) [inline includes/general.php:975]: This is incorrect. Passing $pod would pass the pod from the shortcode, not the actual current Pod.

File: includes/general.php [hunk 8]
```diff
@@ -972,15 +972,15 @@ function pods_shortcode_run( $tags, $content = null ) {
...
 				}
 			}
 
 			if ( 0 < strlen( $tags['having'] ) ) {
 				$params['having'] = $tags['having'];
 
 				if ( pods_shortcode_allow_evaluate_tags() ) {
-					$params['having'] = pods_evaluate_tags( html_entity_decode( $params['having'] ), true, '""', true );
+					$params['having'] = pods_evaluate_tags( html_entity_decode( $params['having'] ), true, '""', $pod );
```
Comment (JoryHogeveen) [inline includes/general.php:983]: This is incorrect. Passing $pod would pass the pod from the shortcode, not the actual current Pod.

File: includes/general.php [hunk 9]
```diff
@@ -926,15 +926,15 @@ function pods_shortcode_run( $tags, $content = null ) {
 			$id = $tags['slug'];
 
 			if ( pods_shortcode_allow_evaluate_tags() ) {
-				$id = pods_evaluate_tags( $id, true, null, true );
+				$id = pods_evaluate_tags( $id, true );
```
Comment (JoryHogeveen) [inline includes/general.php]: You do not want to support pod related tags for `slug`?

File: includes/general.php [hunk 10]
```diff
@@ -926,15 +926,15 @@ function pods_shortcode_run( $tags, $content = null ) {
...
 			}
 		}
 
 		if ( ! empty( $tags['id'] ) ) {
 			$id = $tags['id'];
 
 			if ( pods_shortcode_allow_evaluate_tags() ) {
-				$id = pods_evaluate_tags( $id, true, null, true );
+				$id = pods_evaluate_tags( $id, true );
```
Comment (JoryHogeveen) [inline includes/general.php]: You do not want to support pod related tags for `id`?

**Comment by cweagans** [comment]
I'm _very_ excited about this. I spent most of a day this weekend trying to figure out just the right magic tag to throw into my `where` clause, but it of course does not work yet. Looking forward to having this functionality available for my site!

File: includes/data.php [hunk 11]
```diff
@@ -1479,20 +1490,33 @@ function pods_mb_substr( $string, $start, $length = null, $encoding = null ) {
...
- *
- * @param string|array|object $tags     String to be evaluated.
- * @param bool                $sanitize Whether to sanitize.
- * @param null|mixed          $fallback The fallback value to use if not set, should already be sanitized.
+ * @param string|array|object $tags String to be evaluated.
+ * @param array               $args {
+ *     Function arguments.
+ *     @type bool       $sanitize Whether to sanitize.
+ *     @type null|mixed $fallback The fallback value to use if not set, should already be sanitized.
+ *     @type bool|Pods  $pod      Pod to parse the tags through.
```
[the remark this answers is not in the document]
Comment (JoryHogeveen) [reply includes/data.php:1498]: Bump! :)

[the remark this answers is not in the document]
Comment (sc0ttkclark) [reply includes/data.php:1498]: Instead of passing `true` as `$pod`, can we add another argument `$use_current_pod` that can be used instead?

[the remark this answers is not in the document]
Comment (JoryHogeveen) [reply includes/data.php]: Done!

File: includes/data.php [hunk 12]
```diff
@@ -1653,6 +1740,16 @@ function pods_evaluate_tag( $tag, $sanitize = false, $fallback = null ) {
...
 	} elseif ( 2 === count( $tag ) ) {
 		$value = pods_v( $tag[1], $tag[0], null );
+	} else {
+		// Some magic tags support traversal.
+		$value = pods_v( array_slice( $tag, 1 ), $tag[0], null );
+	}
+
+	if ( $helper ) {
+		if ( ! $pod ) {
+			$pod   = pods();
```
Comment (tr1b0t) [inline includes/data.php:1750]: :warning: Equals sign not aligned correctly; expected 1 space but found 3 spaces

<sub>hash: 8b97d8c2cae78aae0bd92820cb56a996</sub>

File: classes/Pods.php (hunk 1 above)
Comment (JoryHogeveen) [reply classes/Pods.php:4199]: This part is required for this PR to work properly. It's based on PR: #5713 which will be closed automatically once this PR is merged.

File: includes/data.php [hunk 13]
```diff
@@ -1653,6 +1742,16 @@ function pods_evaluate_tag( $tag, $sanitize = false, $fallback = null ) {
 		$value = pods_v( $tag[0], 'get', null );
 	} elseif ( 2 === count( $tag ) ) {
 		$value = pods_v( $tag[1], $tag[0], null );
+	} else {
+		// Some magic tags support traversal.
+		$value = pods_v( array_slice( $tag, 1 ), $tag[0], null );
+	}
+
+	if ( $helper ) {
+		if ( ! $pod instanceof Pods  ) {
```
Comment (tr1b0t) [inline includes/data.php:1751]: :no_entry_sign: Expected exactly one space before closing parenthesis; "  " found.

<sub>hash: 4c60a72eb369ed7eac5516ecf448b2f4</sub>

**Review by sc0ttkclark — DISMISSED** [review]

**Comment by JoryHogeveen** [comment]
@sc0ttkclark 
Since magic tags already have various unit tests, want me to check and add more? Possibly some related to this PR?

File: includes/data.php [hunk 14]
```diff
@@ -1524,20 +1535,34 @@ function pods_mb_substr( $string, $start, $length = null, $encoding = null ) {
...
+ * }
  *
  * @return string
  *
+ * @since 2.1
+ *
  * @see pods_evaluate_tag
  */
-function pods_evaluate_tags( $tags, $sanitize = false, $fallback = null ) {
+function pods_evaluate_tags( $tags, $args = array() ) {
```
Comment (tr1b0t) [inline includes/data.php:1553]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 0798ce6e6917bd1fce0da922565dce2b</sub>

File: includes/data.php [hunk 15]
```diff
@@ -1524,20 +1535,34 @@ function pods_mb_substr( $string, $start, $length = null, $encoding = null ) {
...
+
+	// Back compat.
+	if ( ! is_array( $args ) ) {
+		$prev_args = array( 'tags', 'sanitize', 'fallback' );
```
Comment (tr1b0t) [inline includes/data.php:1557]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 43251be5cca543966fb32a26b670568d</sub>

File: includes/data.php [hunk 16]
```diff
@@ -1547,48 +1572,84 @@ function pods_evaluate_tags( $tags, $sanitize = false, $fallback = null ) {
...
+ *     @type bool       $use_current_pod Whether to auto-detect the current Pod.
+ * }
  *
  * @return string Evaluated content.
  *
  * @see pods_evaluate_tag
  */
-function pods_evaluate_tag_sanitized( $tag ) {
-	return pods_evaluate_tag( $tag, true );
+function pods_evaluate_tag_sanitized( $tag, $args = array() ) {
```
Comment (tr1b0t) [inline includes/data.php:1606]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: f31eac4292d25559c159d9d0aa4d5a2c</sub>

File: includes/data.php [hunk 17]
```diff
@@ -1547,48 +1572,84 @@ function pods_evaluate_tags( $tags, $sanitize = false, $fallback = null ) {
...
+	if ( ! is_array( $args ) ) {
+		$args = array();
```
Comment (tr1b0t) [inline includes/data.php:1608]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 29f49acb234c7401e5581af4fabb15e7</sub>

File: includes/data.php [hunk 18]
```diff
@@ -1547,48 +1572,84 @@ function pods_evaluate_tags( $tags, $sanitize = false, $fallback = null ) {
...
+ *     @type null|mixed $fallback        The fallback value to use if not set, should already be sanitized.
+ *     @type bool|Pods  $pod             Pod to parse the tags through.
+ *     @type bool       $use_current_pod Whether to auto-detect the current Pod.
+ * }
  *
  * @return string Evaluated content.
  */
-function pods_evaluate_tag( $tag, $sanitize = false, $fallback = null ) {
-	global $wpdb;
+function pods_evaluate_tag( $tag, $args = array() ) {
```
Comment (tr1b0t) [inline includes/data.php:1630]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 2fd3788ae9c96779e2b25d161b7e5e07</sub>

File: includes/data.php [hunk 19]
```diff
@@ -1547,48 +1572,84 @@ function pods_evaluate_tags( $tags, $sanitize = false, $fallback = null ) {
...
+	$defaults = array(
```
Comment (tr1b0t) [inline includes/data.php:1631]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: d9eb9bba6b40377075f7db09410d9bcc</sub>

File: includes/data.php [hunk 20]
```diff
@@ -1547,48 +1572,84 @@ function pods_evaluate_tags( $tags, $sanitize = false, $fallback = null ) {
...
+
+	// Back compat.
+	if ( ! is_array( $args ) ) {
+		$args = func_get_args();
+
+		$defaults['sanitize'] = $args[1];
+		if ( isset( $args[2] ) ) {
+			$defaults['fallback'] = $args[2];
+		}
+		$args = array();
```
Comment (tr1b0t) [inline includes/data.php:1646]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 9925f82d5c903723122bc304e6ed4df5</sub>

File: includes/general.php [hunk 21]
```diff
@@ -922,19 +922,25 @@ function pods_shortcode_run( $tags, $content = null ) {
 		// id > slug (if both exist)
 		$id = null;
 
+		$evaluate_tags_args = array(
```
Comment (tr1b0t) [inline includes/general.php:925]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 7b3c4d1783b6ec1fa06ca2b46f8c79a0</sub>

File: includes/general.php [hunk 22]
```diff
@@ -964,6 +970,12 @@ function pods_shortcode_run( $tags, $content = null ) {
 		$params = array();
 
 		if ( ! defined( 'PODS_DISABLE_SHORTCODE_SQL' ) || ! PODS_DISABLE_SHORTCODE_SQL ) {
+			$evaluate_tags_args = array(
```
Comment (tr1b0t) [inline includes/general.php:973]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: ece3b20d843d6bfeace0ea023d96db93</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 9 error(s)

File: includes/data.php (hunk 20 above)
Comment (sc0ttkclark) [inline includes/data.php:1646]:
```suggestion
		$prev_args = array( 'tag', 'sanitize', 'fallback' );
		$args      = func_get_args();
		$args      = array_combine( array_slice( $prev_args, 0, count( $args ) ), $args );
		unset( $args['tag'] );
```

**Review by sc0ttkclark — APPROVED** [review]
